torch/cuda/amp/grad_scaler.py

```diff
@@ -180,25 +180,36 @@ def _unscale_grads_(self, optimizer, inv_scale, found_inf, allow_fp16):
         per_device_inv_scale = _MultiDeviceReplicator(inv_scale)
         per_device_found_inf = _MultiDeviceReplicator(found_inf)
 
+        # To set up _amp_foreach_non_finite_check_and_unscale_, split grads by device and dtype.
+        # There could be hundreds of grads, so we'd like to iterate through them just once.
+        # However, we don't know their devices or dtypes in advance.
+        per_device_and_dtype_grads = defaultdict(defaultdict(list))  # https://knowyourmeme.com/memes/xzibit-yo-dawg
```

Comment:
now HERE'S a comment that can be removed...when i wrote this last night i felt it was illegal to make a defaultdict of a defaultdict without some remark from the peanut gallery

Reply:
Removed :(